contracts/cw721-base/src/state.rs

```diff
@@ -46,7 +46,8 @@ pub fn increment_tokens(storage: &mut dyn Storage) -> StdResult<u64> {
 }
 
 pub struct TokenIndexes<'a> {
-    pub owner: MultiIndex<'a, TokenInfo>,
+    // pk goes to second tuple element
+    pub owner: MultiIndex<'a, (PkOwned, PkOwned), TokenInfo>,
```

Comment:
Do we have to use PkOwned and not &str?

(That can be a future optimization, but curious)

Reply:
We're using `PkOwned` just because `PrimaryKey` is implemented for it. See https://github.com/CosmWasm/cosmwasm-plus/blob/4bbb2050c288bf11b4fe2933cc1f4975d144b94a/packages/storage-plus/src/keys.rs#L153L157.

It would be a matter of implementing `PrimaryKey` for `Vec[u8]` and friends. I can do it, but maybe in a different PR would be a good idea.

Reply:
We also implement PrimaryKey for &[u8] and &str with some lifetime requirements. I *think* this index usage will satisfy the lifetime requirements: https://github.com/CosmWasm/cosmwasm-plus/blob/4bbb2050c288bf11b4fe2933cc1f4975d144b94a/packages/storage-plus/src/keys.rs#L26-L53

The point is not just to avoid a copy, but also to make the usage much simpler. You can just do `&canonical_addr` if the key is `&[u8]` for example

Reply:
Supporting Vec<u8> directly not the PkOwned wrapper is definitely another PR

Reply:
You're right. I'll give it a try.

Reply:
I now know why I used `PkOwned`: To avoid running into lifetime conflicts.

Slices cannot be used, because they don't have size known at compile time. It's the same with `str`.

That leave us with:

- `PkOwned`
- `Vec[u8]`, but it lacks trait impls. So, `PkOwned`.
- `String`. Again, it lacks needed trait impls, and probably for a good reason: unsupported bytes, utf-8 compliance, etc.

So, to move forward, we either:
- Stick to `PkOwned`, and improve usability through `Into` and `From`.
- Implement traits for `Vec[u8]` directly, and remove `PkOwned`. There's [a ticket for that already](#199).
- Fix the lifetime conflicts of `MultiIndex` and related methods, which receive a lifetime as a parameter. This is probably doable, but will require some time / effort.

Reply:
Sounds good.

> * Stick to PkOwned, and improve usability through Into and From.
> * Implement traits for Vec[u8] directly, and remove PkOwned. There's a ticket for that already (#199).

Of these two, I prefer the second. It's in a reasonable place in the backlog. We'll get to it after the major work ahead.

> * Fix the lifetime conflicts of UniqueIndex, MultiIndex and their methods (those which receive a lifetime as a parameter). This is probably doable, but will require some time, and effort.

I made a ticket for this one: #232